module/zfs/dmu_recv.c

```diff
@@ -1243,6 +1300,182 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+	zio_t *io;
+	zbookmark_phys_t zb;
+	dnode_t *dn;
+	abd_t *abd = rrd->abd;
+	zio_cksum_t bp_cksum = bp->blk_cksum;
+	enum zio_flag flags = ZIO_FLAG_SPECULATIVE |
+	    ZIO_FLAG_DONT_CACHE | ZIO_FLAG_DONT_RETRY | ZIO_FLAG_CANFAIL;
+
+	if (rwa->raw)
+		flags |= ZIO_FLAG_RAW;
+
+	err = dnode_hold(rwa->os, drrw->drr_object, FTAG, &dn);
+	if (err != 0)
+		return (err);
+	SET_BOOKMARK(&zb, dmu_objset_id(rwa->os), drrw->drr_object, 0,
+	    dbuf_whichblock(dn, 0, drrw->drr_offset));
+	dnode_rele(dn, FTAG);
+
+	if ((drrw->drr_compressiontype != BP_GET_COMPRESS(bp)) ||
+	    (!rwa->raw && BP_USES_CRYPT(bp))) {
+		if (DRR_WRITE_COMPRESSED(drrw)) {
+			/* Decompress the stream data */
+			abd_t *dabd = abd_alloc_linear(
+			    drrw->drr_logical_size, B_FALSE);
+			err = zio_decompress_data(drrw->drr_compressiontype,
+			    abd, abd_to_buf(dabd), abd_get_size(abd),
+			    abd_get_size(dabd), NULL);
+			if (err != 0) {
+				abd_free(dabd);
+				return (err);
+			}
+			/* Swap in the newly decompressed data into the abd */
+			abd_free(abd);
+			abd = dabd;
+		}
+
+		/*
+		 * The stream is not encrypted but the data on-disk is.
+		 * We need to re-encrypt the buf using the same
+		 * encryption type, salt, iv, and mac that was used to encrypt
+		 * the block previosly.
+		 */
+		if (!rwa->raw && BP_USES_CRYPT(bp)) {
+			dsl_dataset_t *ds;
+			dsl_crypto_key_t *dck = NULL;
+			uint8_t salt[ZIO_DATA_SALT_LEN];
+			uint8_t iv[ZIO_DATA_IV_LEN];
+			uint8_t mac[ZIO_DATA_MAC_LEN];
+			boolean_t no_crypt = B_FALSE;
+			dsl_pool_t *dp = dmu_objset_pool(rwa->os);
+			abd_t *eabd = abd_alloc_linear(drrw->drr_logical_size,
+			    B_FALSE);
+			zio_crypt_decode_params_bp(bp, salt, iv);
+			zio_crypt_decode_mac_bp(bp, mac);
+
+			dsl_pool_config_enter(dp, FTAG);
+			err = dsl_dataset_hold_flags(dp, rwa->tofs,
+			    DS_HOLD_FLAG_DECRYPT, FTAG, &ds);
+			if (err != 0) {
+				dsl_pool_config_exit(dp, FTAG);
+				abd_free(eabd);
+				return (SET_ERROR(EACCES));
+			}
+
+			/* Look up the key from the spa's keystore */
+			err = spa_keystore_lookup_key(rwa->os->os_spa,
+			    zb.zb_objset, FTAG, &dck);
+			if (err != 0) {
+				dsl_dataset_rele_flags(ds, DS_HOLD_FLAG_DECRYPT,
+				    FTAG);
+				dsl_pool_config_exit(dp, FTAG);
+				abd_free(eabd);
+				return (SET_ERROR(EACCES));
+			}
+
+			err = zio_do_crypt_abd(B_TRUE, &dck->dck_key,
+			    BP_GET_TYPE(bp), BP_SHOULD_BYTESWAP(bp), salt, iv,
+			    mac, drrw->drr_logical_size, abd, eabd, &no_crypt);
+
+			spa_keystore_dsl_key_rele(rwa->os->os_spa, dck, FTAG);
+			dsl_dataset_rele_flags(ds, DS_HOLD_FLAG_DECRYPT, FTAG);
+			dsl_pool_config_exit(dp, FTAG);
+
+			ASSERT0(no_crypt);
+			if (err != 0) {
+				abd_free(eabd);
+				return (err);
+			}
+			/* Swap in the newly encrypted data into the abd */
+			abd_free(abd);
+			abd = eabd;
+
+			/*
+			 * We want to prevent zio_rewrite() from trying to
+			 * encrypt the data again
+			 */
+			flags |= ZIO_FLAG_RAW;
+		}
+
+		if (BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF) {
```

Comment:
Probably a silly question, but doesn't this result in compressing the already encrypted data (which is going to not be compressible)? Don't we want the other order?

Reply:
Good point, yeah the order was wrong I've fixed that but it got me wondering why testing didn't catch that. And then I realized that I'm generating incompressible data. Fixing that revealed another bug. Thanks for taking a look